[Feat] #15 - Bar 컴포넌트 UI 구현

kms0233 opened the pull request:
## 🧩 작업 내용
<!-- 작업한 내용을 간단하게 적어주세요! -->
Bar 컴포넌트 구현했습니다아!!

## 🪁 PR Point
<!-- 주의할 사항이나 같이 고민해볼 부분, 강조하고 싶은 내용 등을 적어주세요! -->
ProgressBar는 currentIndex와 totalViews 값을 매개변수로 전달해 사용하시면 됩니당
WeekCategoryBar는 너무 하드 코딩 된거같아서 ㅠㅠ 도와주세염 엄마~~

## 📱 스크린샷
<!-- 작업 내용, 스크린 샷(GIF/MP4) 첨부해주세요. -->
### WeekCategoryBar
<img width="400" alt="image" src="https://github.com/user-attachments/assets/65357b20-d946-41e6-881b-e4f535fdc43b" />

### ProgressBar
<img width="400" alt="image" src="https://github.com/user-attachments/assets/b940c7e1-9cc8-448a-80d1-1bb650b28e61" />

### CategoryBar
<img width="400" alt="image" src="https://github.com/user-attachments/assets/5616be46-7fb1-4187-a603-c1bbde4ee20c" />


### 🔗 Issue 
<!-- 생성한 관련 이슈가 있다면 Resolved #이슈번호로 닫아주세요! -->
Resolved #15

kms0233 committed:
[Fix/#15] 카데고리 enum 수정

kms0233 committed:
[Feat/#15] CategoryBar 컴포넌트 구현

kms0233 committed:
[Feat/#15] ProgressBar 컴포넌트 구현

kms0233 committed:
[Feat/#15] WeekCategoryBar 컴포너트 구현

kms0233 committed:
Merge branch 'develop' into feat/#15-BarComponent

# Conflicts:
#	Gongbaek_iOS/Gongbaek_iOS/Global/Component/Button/CategoryImageButton.swift

kms0233 committed:
[Fix/#15] conflict 해결

kms0233 committed:
Merge branch 'develop' into feat/#15-BarComponent

kms0233 committed:
Merge branch 'develop' into feat/#15-BarComponent

kms0233 committed:
Merge branch 'develop' into feat/#15-BarComponent

HEHEEUN commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Bar/ProgressBar.swift:
```diff
@@ -0,0 +1,43 @@
+        }
+        .frame(height: 7)
+    }
+}
```
직접 짠 거 완전 기특해 ! 이 실력이 어떻게 02?

Nya128 replied:
김민서는 사실 유명한 02 갓기임

HEHEEUN reviewed (approved):
고생했다 공이ㄹ민서!

Nya128 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Bar/WeekCategoryBar.swift:
```diff
@@ -0,0 +1,73 @@
+    case fri = "금요일"
+}
+
+struct WeekCategoryBar: View {
```
네이밍을 수정하면 좋을 거 같은데... 흠
'카테고리'라는 이름이 여기저기 자주 쓰이니까 헷갈릴 수도 있어서
차라리 요일 선택바, 요일 필터바 어떤가요?! 🤣 
WeekSelectionBar, WeekFilterBar 중에 괜찮아보이는걸로 바꿔주심 됩니당ㅎ

Nya128 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Bar/CategoryBar.swift:
```diff
@@ -0,0 +1,32 @@
+import SwiftUI
+
+struct CategoryBar: View {
+    @State private var selectedCategory: CategoryState = .all
```
기본값 넣어둔거 좋다~

Nya128 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Bar/CategoryBar.swift:
```diff
@@ -0,0 +1,32 @@
+    @State private var selectedCategory: CategoryState = .all
+
+    var body: some View {
+        ScrollView(.horizontal, showsIndicators: false) {
+            HStack(spacing: 6) {
+                ForEach(CategoryState.allCases, id: \.self) { category in
+                    CategoryButton(
+                        text: category.categoryName,
+                        isSelcted: selectedCategory == category
+                    ) {
+                        selectedCategory = category
+                    }
+                }
+            }
+            .padding(.horizontal, 16)
+        }
```
난 여기서 스유의 힘을 느꼈어... 이렇게 짧게 스크롤뷰 작성이 가능하다니. 각 버튼마다 액션 붙이는게 이렇게 쉽다니. 🙂‍↔️

Nya128 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Bar/ProgressBar.swift:
```diff
@@ -0,0 +1,43 @@
+
+struct ProgressBar: View {
+    @Binding var currentIndex: Int
+    var totalViews: Int
```
- Int값 변수라면 number나 count가 들어가면 이해하기 쉬울듯!!
`totalNumberOfViews`가 더 적합한 네이밍인듯 함다
- 근데 totalViews 개수를 생성자 파라미터로 직접 넘기는 것보다,, 이것도 enum 만들어서 ProgressBar 사용되는 뷰가 어딘지를 넘기는 게 편할 것 같긴해욤 
그러면 enum에 totalViews 개수를 우리가 모아서 통일성있게 관리할 수 있어서?!
```
enum ProgressBarViewType: Int {
    case onboarding = 8
    case addMeeting = 8
}
```

kms0233 replied:
네 엄마!! 히히

Nya128 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Bar/ProgressBar.swift:
```diff
@@ -0,0 +1,43 @@
+struct ProgressBar: View {
+    @Binding var currentIndex: Int
+    var totalViews: Int
+    var progressRatio: CGFloat {
+        totalViews > 0 ? CGFloat(currentIndex + 1) / CGFloat(totalViews) : 0
+    }
```
여기서 Swift 고수 느낌 솔솔 ㅜ 프로퍼티 자유자재로 막 다룸

Nya128 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Bar/ProgressBar.swift:
```diff
@@ -0,0 +1,43 @@
+                        width: geometry.size.width * progressRatio,
+                        height: 7
+                    )
+                    .animation(.easeInOut, value: currentIndex)
```
애니메이션 효과까지 넣는 꼼꼼함은 어딜가면 배울 수 있나요?

Nya128 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Bar/ProgressBar.swift:
```diff
@@ -0,0 +1,43 @@
+                Rectangle()
+                    .fill(.gray10)
+                    .frame(
+                        width: geometry.size.width * progressRatio,
```
GeometryReader를 적재적소에 쓸줄아는 민서공주

Nya128 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Bar/WeekCategoryBar.swift:
```diff
@@ -0,0 +1,73 @@
+        .frame(height: 38)
+        .padding(.horizontal, 3)
+    }
+
```
공백 삭제!

Nya128 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Bar/WeekCategoryBar.swift:
```diff
@@ -0,0 +1,73 @@
+        let buttonWidth = geometry.size.width / CGFloat(Day.allCases.count)
+        return CGFloat(index) * buttonWidth - geometry.size.width / 2 + buttonWidth / 2
+    }
+    
```
여기두 공백 삭제!

Nya128 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Bar/WeekCategoryBar.swift:
```diff
@@ -0,0 +1,73 @@
+    }
+    
+    func sliderOffset(in geometry: GeometryProxy) -> CGFloat {
+        let index = Day.allCases.firstIndex(of: selectedDay) ?? 0
```
여기는 디폴트값을 설정해주기보다는 guard let 사용하면 더 깔끔해질 듯?

Nya128 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Bar/WeekCategoryBar.swift:
```diff
@@ -0,0 +1,73 @@
+                RoundedRectangle(cornerRadius: 4)
+                    .fill(.gray01)
+                
+                RoundedRectangle(cornerRadius: 2)
+                    .fill(.white)
+                    .frame(width: buttonWidth(in: geometry) - 6, height: 34)
+                    .offset(x: sliderOffset(in: geometry))
+                    .animation(.easeInOut(duration: 0.2), value: selectedDay)
+                
+                HStack(spacing: 0) {
+                    ForEach(Day.allCases, id: \.self) { day in
+                        Button(action: {
+                            withAnimation {
+                                selectedDay = day
+                            }
+                        }) {
+                            Text(day.rawValue)
+                                .font(.pretendard(.caption1_sb_13))
+                                .frame(maxWidth: .infinity)
+                                .frame(height: 36)
+                                .foregroundColor(selectedDay == day ? .mainOrange : .gray06)
+                        }
+                    }
+                }
```
선택 효과를 RoundedRectangle로 하지말고 Button 백그라운드로 색상을 넣으면 더 좋을 거 같긴한데...!! 리팩토링 한번 해볼게여

kms0233 replied:
ㅠㅠㅠ 엄마가 없으면 난 앱잼을 못해

Nya128 reviewed (commented):
ProgressBar 비율 넣어서 한거 완벽.
WeekCategoryBar도 세그먼트 컨트롤 직접 만든거 대단해요
하드코딩한 것 같다는 부분 같이 고민해봤는데... RoundedRectangle 빼고 Button으로만 구현하는 방향으로 수정했고,
디자인과 최대한 비슷하게 구현해보고 싶어서 😇 전체 버튼 좌우 패딩이 다른 요일 버튼보다 2만큼 작은 것까지 고려해서 너비 맞춰봤어용

Nya128 committed:
[Refactor/#15] 채우기 요일바 리팩토링

- 전체 버튼 내부 패딩이 요일 버튼보다 좌우 2씩 작아서 모든 버튼 너비를 균등하게 나눌 경우 4만큼의 공간이 남는 문제 해결
- RoundedRectangle로 구현한 선택 효과 삭제 후 Button Background로 효과 적용

Nya128 commented:
내가 요일 바 수정한 코드 올렸는데 한번 확인해보구 이해 안 가는 거 있음 말해주🤤
남은 리뷰만 수정해서 반영해주면 됩니다~~❤️

kms0233 committed:
[Fix/#15] 코드 리뷰 반영
